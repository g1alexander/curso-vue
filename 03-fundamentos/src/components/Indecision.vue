<template>
  <img v-if="img" :src="img" alt="bg" />
  <div class="bg-dark"></div>
  <div class="indecision-container">
    <input v-model="question" type="text" placeholder="Hazme una pregunta" />
    <p>Recuerda terminar con un signo de ?</p>

    <div v-if="isValid">
      <h2>{{ question }}</h2>
      <h1>{{ answer === "yes" ? "Si!" : "No!" }}</h1>
    </div>
  </div>
</template>

<script>
export default {
  data() {
    return {
      question: "",
      img: null,
      answer: null,
      isValid: false,
    };
  },
  methods: {
    async getWtf() {
      try {
        this.answer = "...pensando";
        const response = await fetch("https://yesno.wtf/api");
        const { answer, image } = await response.json();

        this.answer = answer;
        this.img = image;
      } catch (error) {
        this.answer = "no se pudo cargar el api";
        this.img = null;
        console.error(error);
      }
    },
  },
  watch: {
    question(value, oldValue) {
      // console.log({ value, oldValue });

      this.isValid = false;

      console.log({ value});

      if (!value.includes("?")) return;

      this.isValid = true;

      this.getWtf();
    },
  },
};
</script>

<style scoped>
img,
.bg-dark {
  height: 100vh;
  left: 0px;
  max-height: 100%;
  max-width: 100%;
  position: fixed;
  top: 0px;
  width: 100vw;
}

.bg-dark {
  background-color: rgba(0, 0, 0, 0.4);
}

.indecision-container {
  position: relative;
  z-index: 99;
}

input {
  width: 250px;
  padding: 10px 15px;
  border-radius: 5px;
  border: none;
}
input:focus {
  outline: none;
}

p {
  color: white;
  font-size: 20px;
  margin-top: 0px;
}

h1,
h2 {
  color: white;
}

h2 {
  margin-top: 150px;
}
</style>
