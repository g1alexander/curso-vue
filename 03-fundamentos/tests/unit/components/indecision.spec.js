import Indecision from "@/components/Indecision.vue"
import { shallowMount } from "@vue/test-utils"

describe('componente Indecision.vue', () => {
  let wrapper = shallowMount(Indecision);
  let clgSpy;
  let getWtfSpy;
  global.fetch = jest.fn(() => Promise.resolve({
    json: () => Promise.resolve({
      "answer": "yes",
      "forced": false,
      "image": "https://yesno.wtf/assets/yes/2.gif"
    })
  }));

  beforeEach(() => {
    wrapper = shallowMount(Indecision);
    clgSpy = jest.spyOn(console, 'log');
    getWtfSpy = jest.spyOn(wrapper.vm, 'getWtf'); //.vm accede al script de vue

    jest.clearAllMocks() //limpia mock por cada test
  })

  
  test('match con snapshot', () => {
    expect(wrapper.html()).toMatchSnapshot()
  });


  test('escribir en el input no debe disparar nada (console.log)', async () => {

    const $input = wrapper.find("input");
    await $input.setValue("hola mundo")

    expect(clgSpy).toHaveBeenCalled()
    expect(getWtfSpy).not.toHaveBeenCalled()
    
  });
  
  test('escribir simbolo de interrogacion dispara getWtf', async () => {
    const $input = wrapper.find("input");
    await $input.setValue("hola?");
    
    expect(clgSpy).toHaveBeenCalled()
    expect(getWtfSpy).toHaveBeenCalled();
  });

  test('pruebas getWtf', async () => {
    await wrapper.vm.getWtf();
    
    const $img = wrapper.find("img");
    
    expect($img).toBeTruthy()
    expect(wrapper.vm.img).toBe("https://yesno.wtf/assets/yes/2.gif");
    expect(wrapper.vm.answer).toBe("yes")
  });
  
  test('fallo en el api', async () => {
    fetch.mockImplementationOnce(() => Promise.reject("API is down"));
    
    await wrapper.vm.getWtf();
    const $img = wrapper.find("img");
    expect($img.exists()).toBeFalsy()
    expect(wrapper.vm.answer).toBe("no se pudo cargar el api")
  });
});