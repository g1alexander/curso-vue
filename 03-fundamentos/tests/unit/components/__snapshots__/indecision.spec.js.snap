// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`componente Indecision.vue match con snapshot 1`] = `
<!--v-if-->
<div class="bg-dark"></div>
<div class="indecision-container"><input type="text" placeholder="Hazme una pregunta">
  <p>Recuerda terminar con un signo de ?</p>
  <!--v-if-->
</div>
`;
